### DISCUSSION TITLE
Clarification on array in FPP

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#2082] [Q&A]
In FPP I can create an array in two way:
`my_array    : [10] F32`
or
`array my_array = [20] F32`

What's the design choice behind? When I should use one or the other? 

Moreover it seems that the first approach is not supported by fprime-GDS, is it planning to be supported in next versions?

**Comment by thomas-bc** [comment]
I don't think they are meant to be interchangeable. From the [FPP design spec](https://fprime-community.github.io/fpp/fpp-spec.html):
- An [array definition](https://fprime-community.github.io/fpp/fpp-spec.html#Definitions_Array-Definitions) would be  `array my_array = [20] F32`
- The `my_array : [10] F32` syntax is only valid in specific cases such as [members in a struct definition](https://fprime-community.github.io/fpp/fpp-spec.html#Definitions_Struct-Definitions), port arrays etc...

**Reply by SMorettini** [reply]
Thank you for the answer. Do you know if the `my_array : [10] F32` will be supported by fprime-gds? Currently when I use it I just saw the first element and not the full array.

**Reply by thomas-bc** [reply]
I'm not sure I understand the question quite right, could you clarify? Or provide an example of where you're using that syntax?

**Reply by SMorettini** [reply]
Let's say my component contain the following code:
```
...
    array F32x5 = [5] F32
    struct Type_1{
        my_data : F32x5
    }

    struct Type_2{
        my_data :  [5] F32
    }

 ...

    telemetry CHANNEL_1: Type_1 id 1

    telemetry CHANNEL_2: Type_2 id 2
```
In the interface of `fprime-gds`, the telemetry `CHANNEL_1` is shown in the interface but `CHANNEL_2` no.  I think the error come from the dictionary generated: the entry generated for `Type_2` is
```
    <serializable type="Type_2">
      <members>
        <member name="my_data" format_specifier="%g" description="..." default="0.0f" type="F32"/>
      </members>
    </serializable>
```
The question is: is that expected because the array definition in `Type_2` is just for port usage or is a bug in the dictionary generation?

**Reply by thomas-bc** [reply]
@bocchino any thoughts here?

**Reply by bocchino** [reply]
It looks like the Python dictionary generator is not handling member arrays of structs in the expected way. This appears to be one of those mismatches between the FPP spec and the GDS behavior. I'll add a note on this in the FPP User's Guide.

**Reply by bocchino** [reply]
Also, it's not entirely clear what the GDS should do with a member array in a struct, since the array could be large, and not feasible to display member by member. Eventually the GDS should have a spec that describes its behavior in all these cases. Right now these corner cases are handled on a best effort basis.

**Reply by SMorettini** [reply]
> Also, it's not entirely clear what the GDS should do with a member array in a struct, since the array could be large, and not feasible to display member by member.

Gds already support `array my_array = [20] F32`. I have an array of size 2 containing for each element an array of size 200. Gds already show this data. It's not user-friendly but I don't think there would be any problem showing a struct containing a big array, I would expect it to work already. I think the missing part is only to create the dictionary accordingly so Gds know how to interpret the data.

**Reply by bocchino** [reply]
That makes sense for array types, which are limited to size 256. Struct member arrays are not so limited. They are serializable values, so they need to fit on the stack, and so they can't be huge on the order of megabytes. But in practice they could be maybe 1K in size. @LeStarch, @thomas-bc, @timcanham what do you think?

**Reply by LeStarch** [reply]
It depends on the intended usage.  Items flowing to telemetry, events, channels do go on the stack and thus should be kept small.  The big reason is the must fit in a `Fw::ComBuffer`, whose size is globally determined and used for everything.

I can envision a use case where a serializable type is used to pass data through ports, and is substantially bigger and placed in a allocated `Fw::Buffer` for example.
